### PR TITLE
DNM: Use Prost instead off rust-protobuf

### PR DESCRIPTION
**Pull request by nrc**
This is not ready to merge yet because of the `path` dependencies. Once PRs to upstream crates get merged, I can change these to Git or version deps. I'm posting earlier so reviewers can get started looking at it.

## What have you changed? (mandatory)

Changed all uses of protobufs to use Prost rather than rust-protobuf for coding and decoding.

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

Existing tests.

## Does this PR affect documentation (docs) or release note? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Benchmark result if necessary (optional)

tl;dr, we see a small, but consistent improvement (1-2%) across all benchmarks in sysbench and YCSB. Peak latency seems to decrease significantly, but the actual amount is hard to quantify (0-12%, with a median around 4%, but lots of noise).

### CODE REVIEW

**Comment by brson** [comment]
Heroic work pulling all these pieces together and turning them into perf wins @nrc .

File: components/test_coprocessor/src/dag.rs [hunk 1]
```diff
@@ -135,22 +131,22 @@ impl DAGSelect {
...
+        self.aggr_col(col, ExprType::AggBitXor)
     }
 
     pub fn group_by(mut self, cols: &[&Column]) -> DAGSelect {
         for col in cols {
             let offset = offset_for_column(&self.cols, col.id);
-            let mut expr = Expr::new();
-            expr.set_tp(ExprType::ColumnRef);
+            let mut expr = Expr::default();
+            expr.set_tp_(ExprType::ColumnRef);
```
[the remark this answers is not in the document]
Comment (nrc) [reply components/test_coprocessor/src/dag.rs]: Maybe, maybe not - it is because Prost already generates some setters with slightly different semantics so we need to avoid the name clash.

File: src/coprocessor/endpoint.rs [hunk 2]
```diff
@@ -32,9 +31,6 @@ pub struct Endpoint<E: Engine> {
     /// The thread pool to run Coprocessor requests.
     read_pool: ReadPool,
 
-    /// The recursion limit when parsing Coprocessor Protobuf requests.
-    recursion_limit: u32,
```
[the remark this answers is not in the document]
Comment (nrc) [reply src/coprocessor/endpoint.rs]: We're not, sadly. It's just that we can no longer customise the recursion limit (which I think is fine, afaict it is only customised for testing).

File: src/server/status_server.rs [hunk 3]
```diff
@@ -355,7 +355,6 @@ mod tests {
     use crate::config::TiKvConfig;
     use crate::server::status_server::StatusServer;
     use futures::future::{lazy, Future};
-    #[cfg(not(feature = "no-fail"))]
```
[the remark this answers is not in the document]
Comment (nrc) [reply src/server/status_server.rs]: The trait we're including is needed whether or not this feature is on. I don't understand how master is working right now.

File: src/storage/kv/cursor_builder.rs [hunk 4]
```diff
@@ -63,7 +63,7 @@ impl<'a, S: 'a + Snapshot> CursorBuilder<'a, S> {
     /// Set iterator range by giving lower and upper bound.
     /// The range is left closed right open.
     ///
-    /// Both defaults to `None`.
+    /// Both new_s to `None`.
```
[the remark this answers is not in the document]
Comment (nrc) [reply src/storage/kv/cursor_builder.rs]: Bad find and replace, will fix.

File: components/cop_datatype/src/def/field_type.rs [hunk 5]
```diff
@@ -17,6 +17,7 @@ use num_traits::FromPrimitive;
 ///
 /// Please refer to `mysql/type.go` in TiDB.
 #[derive(Primitive, PartialEq, Debug, Clone, Copy)]
+#[repr(i32)]
```
[the remark this answers is not in the document]
Comment (nrc) [reply components/cop_datatype/src/def/field_type.rs]: I think it used to be necessary but no longer is (in an earlier version of the branch). I've removed it.

File: src/raftstore/store/peer_storage.rs [hunk 6]
```diff
@@ -122,7 +123,7 @@ impl EntryCache {
             .take_while(|e| {
                 let cur_idx = end_idx as u64 + cache_low;
                 assert_eq!(e.get_index(), cur_idx);
-                let m = u64::from(e.compute_size());
+                let m = u64::try_from(e.encoded_len()).unwrap();
```
[the remark this answers is not in the document]
Comment (nrc) [reply src/raftstore/store/peer_storage.rs]: It can only panic if we have 128bit pointers. I added a comment.

**Comment by nrc** [comment]
Addressed @brson's comments

**Comment by nrc** [comment]
Closing since the approach has changed somewhat
